### PR TITLE
Use cub::ReduceByKey to count partition indices

### PR DESCRIPTION
**Pull request by codrut3**
This implements a suggestion made by @ekelsen in the comments for #13905.
It replaces the previously custom-made counting method, and is likely more efficient.

In order to use cub::ReduceByKey properly, I defined a specialization of TransformOutputIterator that only allows writes in a bounded interval. This is needed in the case of wrong inputs.

I've also added tests for the GPU kernel, covering the case of wrong inputs.

### CODE REVIEW

File: tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc [hunk 1]
```diff
@@ -57,34 +61,15 @@ __global__ void RangeInitKernel(const T start, const T delta, const int32 size,
...
-// by subtracting the endpoints.
-__global__ void IntervalLengthKernel(int32* start, int32 size, int32* end) {
-  CUDA_1D_KERNEL_LOOP(i, size) {
-    int32 start_point = ldg(start + i);
-    end[i] = end[i] - start_point;
+__global__ void MoveValuesKernel(const int32* keys, const int32* values,
+                                 const int32* size, int32 out_size,
+                                 int32* out) {
+  int32 N = ldg(size);
+  if (N > out_size) N = out_size;
```
Comment (ekelsen) [inline tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc]: int32 N = min(ldg(size), out_size);

File: tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc [hunk 2]
```diff
@@ -130,10 +110,81 @@ void CallGatherKernel(const GPUDevice& d, const T* params, const int32* indices,
       out_size);
 }
 
+struct CubReduceAdd {
```
Comment (ekelsen) [inline tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc]: Can just use cub::Sum()

https://nvlabs.github.io/cub/structcub_1_1_sum.html

**Review by ekelsen — COMMENTED** [review]
Thanks for this nice update!

**Comment by codrut3** [comment]
Hi @ekelsen,

Thank you for doing the review! I implemented the suggestions you made.

**Review by ekelsen — APPROVED** [review]

**Pull request merged**